pkg/admission/admission.go

```diff
@@ -0,0 +1,112 @@
+package admission
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// Config configures a Controller.
+type Config struct {
+	Limit uint64
+}
+
+// Default config settings:
+const (
+	defaultConcurrency = 50
+)
+
+// DefaultConfig returns a Config with preset values.
+func DefaultConfig() Config {
+	c := Config{
+		Limit: defaultConcurrency,
+	}
+
+	return c
+}
+
+// Controller determines the current admission parameters.
+type Controller struct {
+	Pool *quotapool.IntPool
+
+	mu struct {
+		syncutil.RWMutex
+
+		numAcquired int64
+		numWaiting  int64
+	}
+}
+
+// NewController constructs a new Controller struct from a given Config.
+func NewController(conf Config) *Controller {
+	fmt.Println("initializing new controller")
+	c := &Controller{
+
+		// Should this have a better name?
+		Pool: quotapool.NewIntPool("controller intpool", conf.Limit),
+	}
+
+	// Run some goroutine w/ stopper to log stats somewhere.
+	fmt.Printf("intpool cap: %d\n", c.Pool.Capacity())
+	return c
+}
+
+// NumWaiting returns the current number of requests that have been intercepted
+// by admission.Interceptor but have not been acquired.
+func (c *Controller) NumWaiting() int64 {
+	return atomic.LoadInt64(&c.mu.numWaiting)
+}
+
+// NumAcquired returns the number of requests we have acquired quota for.
+// This is functionally identical to computing
+// c.pool.Capacity() - c.pool.ApproximateQuota().
+func (c *Controller) NumAcquired() int64 {
+	return atomic.LoadInt64(&c.mu.numAcquired)
+}
+
+func (c *Controller) handleRelease(alloc *quotapool.IntAlloc) {
+	atomic.AddInt64(&c.mu.numAcquired, -1)
+	c.Pool.Release(alloc)
+}
+
+// Interceptor returns a UnaryServerInterceptor with parameters
+// taken from a Controller.
+func Interceptor(c *Controller) grpc.UnaryServerInterceptor {
+
+	admissionInterceptor := func(ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler) (interface{}, error) {
+
+		atomic.AddInt64(&c.mu.numWaiting, 1)
+
+		if info.FullMethod == "/cockroach.roachpb.Internal/Batch" {
+
+			// Attempt to acquire a single unit from the IntPool.
+			// We're not going to handle the error for now.
+			alloc, _ := c.Pool.Acquire(ctx, 1)
+			atomic.AddInt64(&c.mu.numAcquired, 1)
+			atomic.AddInt64(&c.mu.numWaiting, -1)
+			defer c.handleRelease(alloc)
+
+			// Take request timestamp, add it to context's metadata, and return
+			// this new context.
+			start := time.Now().UnixNano()
+			ctx = metadata.AppendToOutgoingContext(ctx, "start-time", strconv.Itoa(int(start)))
```

Comment:
I don't think this is doing anything for you. You are never reading this and it's probably a tad costly. When we talked about grabbing a timestamp, it was about doing it pre deserialization. This is post deserialization.

Reply:
You're correct - this doesn't really do anything. I removed this section, but what exactly should I be modifying to add a timestamp before deserialization? I incorporated the below suggestion about typechecking `req` as `roachpb.BatchRequest`. Should I be modifying `req.Header` somehow?

Reply:
I think the following will expose the information you want. The stats handler TagRPC method gets called before the body of the request has been read off of the wire (in theory). 

```go
--- a/pkg/rpc/stats_handler.go
+++ b/pkg/rpc/stats_handler.go
@@ -13,6 +13,7 @@ package rpc
 import (
        "context"
        "sync/atomic"
+       "time"
 
        "github.com/cockroachdb/cockroach/pkg/util/log"
        "golang.org/x/sync/syncmap"
@@ -113,17 +114,34 @@ func (sh *StatsHandler) newClient(target string) stats.Handler {
        }
 }
 
+type rpcTimestampKey struct{}
+
+func ContextWithBatchRPCStartTimestamp(ctx context.Context, ts time.Time) context.Context {
+       return context.WithValue(ctx, rpcTimestampKey{}, ts.UnixNano())
+}
+
+func BatchRPCStartTimestampFromContext(ctx context.Context) (_ time.Time, ok bool) {
+       v, ok := ctx.Value(rpcTimestampKey{}).(uint64)
+       if !ok {
+               return time.Time{}, false
+       }
+       return time.Unix(0, v), true
+}
+
 // TagRPC implements the grpc.stats.Handler interface. This
 // interface is used directly for server-side stats recording.
 func (sh *StatsHandler) TagRPC(ctx context.Context, rti *stats.RPCTagInfo) context.Context {
-       return ctx
+       if rti.FullMethodName != "/cockroach.roachpb.Internal/Batch" {
+               return ctx
+       }
+       return ContextWithBatchRPCStartTimestamp(ctx, time.Now())
 }

```